### ISSUE TITLE
AnimationPlayer waits for 2 physics frames before starting

### ISSUE THREAD

**Issue by securas** [godotengine/godot#29187]
**Godot version:**
3.1.1 Stable

**OS/device including version:**
Windows 10

**Issue description:**
AnimationPlayer running in physics process takes 2 frames before starting after call to play.
So calling play to start an animation implies waiting 2 frames before the animation actually begins. This is a problem for pixel art where every frame counts and a jump animation starts when the character is already mid-air.

The project attached starts counting frames after a timer triggers the start of an animation. The animation only prints a message on the console but the message should show at frame 0, no delay, or at frame 1, assuming a single frame delay as in Godot 3.0.6. But shows only at frame 2.

[test_animation_player.zip](https://github.com/godotengine/godot/files/3220171/test_animation_player.zip)

**Comment by avencherus** [comment]
I found it to be one frame and not two.  It seems like everything at time zero happens on the following frame.

I've modified your test to be more accurate.
[test_animation_player.zip](https://github.com/godotengine/godot/files/3220280/test_animation_player.zip)

**Comment by securas** [comment]
Thanks for checking it out @avencherus 
I think that even though it is one frame, it runs after the node2d physics process and perhaps that is why my tick counter says 2 frames. That is actually what counts for me since I will then have to wait 2 frames before starting the (e.g.) actual jump, in order to synchronize it with the animation.

What puzzled me the most is that I did not have this behavior in 3.0.6. I did not test 2.1.x.

**Comment by avencherus** [comment]
No problem.  There has been a fair number of changes to the AnimationPlayer in 3.1, and very many new bugs have been lingering about unaddressed.

Yeah, you'd have to track that kind of tick counter very carefully since the code is fragmented and executing at different points.  I generally stick to the scene tree's frame counter.

A partial work around here is to use `$anim.seek(0.0, true)` to get the animation keyframes to happen immediately.  I say partial, because _it doesn't execute the function call keyframe._

I tested in 3.0.6, and the transformation keyframes do go immediately, however, the call keyframe still lags behind by 1 frame.

**Comment by securas** [comment]
I'm trying to use animationtree, which adds another layer of complexity to this issue. So I guess for now I'll stick to traditional animationplayer, no critical method calls or loading textures and try to start the animation one frame ahead of the action. Thanks for the 'seek' workaround... that will help!

**Comment by KoBeWi** [comment]
This is intended, docs mention this behavior in 3.2 (check the `play()` method description).

**Comment by securas** [comment]
> This is intended, docs mention this behavior in 3.2 (check the `play()` method description).

Actually... The docs did not mention this at the time the issue was raised. Calling it as a documentation issue is an unfortunate habit.

**Comment by securas** [comment]
I may be mistaken but I think that I raised the issue because the behavior
was not present in earlier Godot versions (2.x)

**Comment by KoBeWi** [comment]
Well, it was added for a reason. The old behavior was probably causing problems. Maybe we could make it optional, but it depends if it's possible to warn about potential problems of immediate playing (i.e. if something behaves unexpectedly, the user should be informed about it, so that they can use deferred playing. Similarly to various properties of physics objects).

**Comment by elvisish** [comment]
I’m having issues with a oneshot node delayed by a frame or two, running inside an animationtree, is there a fix for getting it to start immediately?
